src/librustc_mir/interpret/util.rs

```diff
@@ -0,0 +1,73 @@
+use rustc_middle::mir::interpret::InterpResult;
+use rustc_middle::ty::{self, Ty, TyCtxt, TypeFoldable, TypeVisitor};
+use std::convert::TryInto;
+
+/// Returns `true` if a used generic parameter requires substitution.
+crate fn ensure_monomorphic_enough<'tcx, T>(tcx: TyCtxt<'tcx>, ty: T) -> InterpResult<'tcx>
+where
+    T: TypeFoldable<'tcx>,
+{
+    debug!("ensure_monomorphic_enough: ty={:?}", ty);
+    if !ty.needs_subst() {
+        return Ok(());
+    }
+
+    struct UsedParamsNeedSubstVisitor<'tcx> {
+        tcx: TyCtxt<'tcx>,
+    };
+
+    impl<'tcx> TypeVisitor<'tcx> for UsedParamsNeedSubstVisitor<'tcx> {
+        fn visit_const(&mut self, c: &'tcx ty::Const<'tcx>) -> bool {
+            if !c.needs_subst() {
+                return false;
+            }
+
+            match c.val {
+                ty::ConstKind::Param(..) => true,
+                _ => c.super_visit_with(self),
+            }
+        }
+
+        fn visit_ty(&mut self, ty: Ty<'tcx>) -> bool {
+            if !ty.needs_subst() {
+                return false;
+            }
+
+            match ty.kind {
+                ty::Param(_) => true,
+                ty::Closure(def_id, substs)
+                | ty::Generator(def_id, substs, ..)
+                | ty::FnDef(def_id, substs) => {
+                    let unused_params = self.tcx.unused_generic_params(def_id);
+                    for (index, subst) in substs.into_iter().enumerate() {
+                        let index = index
+                            .try_into()
+                            .expect("more generic parameters than can fit into a `u32`");
+                        let is_used =
+                            unused_params.contains(index).map(|unused| !unused).unwrap_or(true);
+                        // Only recurse when generic parameters in fns, closures and generators
+                        // are used and require substitution.
+                        if is_used && subst.needs_subst() {
+                            // Just in case there are closures or generators within this subst,
+                            // recurse.
+                            if subst.super_visit_with(self) {
+                                // Only return when we find a parameter so the remaining substs
+                                // are not skipped.
+                                return true;
+                            }
+                        }
```

Comment:
```suggestion
                        } else if cfg!(debug_assert)
                             match subst.unpack() {
                                 ty::GenericArgKind::Lifetime(_) => ??? // not sure how lt are used
                                 ty::GenericArgKind::Ty(ty) => assert!(matches!(ty.kind, ty::Param(_))),
                                 ty::GenericArgKind::Const(ct) => assert!(matches!(ct, ty::ConstKind::Param(_))),
                             }
                        }
```
Or something like this would make me more comfortable about `type_id`

Reply:
It isn't true that a `subst` here will be a `ty::Param(_)` and not a type which contains a `ty::Param(_)`, right?

Reply:
If a param is unused it, polymorphization should it with `ty::Param(_)`, shouldn't it?

So if this assert failed, we may have called `type_id` with an unpolymorphized type which can result in `type_id::<Ty>() != ty_id::<Ty>()` if we then call `type_id` of the polymorphized type

Reply:
I don't think this could just be an `else` on that conditional, given that if it is used, but doesn't need substitution, then it won't be `ty::Param`, but that assert will be hit regardless. I wasn't sure initially given that `subst` might not be a `ty::Param` but something containing one -  though I think I've convinced myself something like this would probably work.

Reply:
Implemented this in #75155.